### ISSUE TITLE
pykmip-server fails to verify client certificate

### ISSUE THREAD

**Issue by FailedFace** [OpenKMIP/PyKMIP#453]
The following is the error in the server log:
> 2018-09-05 11:32:18,188 - kmip.server - INFO - Dedicating session 00000015 to 10.25.140.176:60669
> 2018-09-05 11:32:18,189 - kmip.server.session.00000015 - INFO - Starting session: 00000015
> 2018-09-05 11:32:18,211 - kmip.server.session.00000015 - DEBUG - Possible session ciphers: 69
> ...
> ...
> ...
> 2018-09-05 11:32:18,224 - kmip.server.session.00000015 - DEBUG - Session cipher selected: ('AES256-SHA', 'TLSv1/SSLv3', 256)
> **2018-09-05 11:32:18,224 - kmip.server.session.00000015 - WARNING - Failure verifying the client certificate.**
> 2018-09-05 11:32:18,225 - kmip.server.session.00000015 - INFO - Stopping session: 00000015

Basically I got a Certificate Signing Request (CSR) from the client and sent it to my server and signed it and created a Signed Certificate. The client then uses the newly signed certificate and the server's certificate to register itself with the kmip server. This seems to fail with this message.

Any ideas on where to go next? I've taken a look at the signed certificate and it's signed with the root CA I'm using with the kmip server. Perhaps there is more information on this WARNING (more like a failure) message? Or maybe I can have the server ignore this check temporarily to investigate the entire flow?

**Comment by PeterHamilton** [comment]
Hi @FailedFace, sorry to hear that you're still hitting certificate issues.

That warning could indicate a couple of things and is generated when one of the following events happens:

1. The client certificate cannot be obtained from the incoming connection.
2. The `ExtendedKeyUsage` extension is not included in the client certificate.
3. The `ExtendedKeyUsage` extension **is** included in the client certificate but is not marked for client authentication.

(1) seems doubtful given you've gotten the client/server working together in the past. (2) is possible; you'll need to use `openssl` to inspect your certificates to see if it has the `ExtendedKeyUsage` extension included. (3) is the same as (2), you'll need to check if the `ExtendedKeyUsage` extension is marked for client authentication.

You can do a quick check for (2) and (3) by setting `enable_tls_client_auth` to `False` in the server config file. This will allow the server to skip checking for the `ExtendedKeyUsage` extension. Restart the server once you've changed the config setting and try using the client again. If things work, that's the problem.

If that doesn't work, we'll need to do a deeper dive on how you're creating your certificates.

**Comment by FailedFace** [comment]
Thanks! Disabling the client auth allowed the key to be served up by the kmip server.
The certificate was generated via: openssl x509 -req -days 365 -in DAMC001-2_CSRFile.csr -CA certs/root_certificate.pem -CAkey certs/root_key.pem -set_serial 0123456789 -out DAMC001-2_CSRFile.crt

Does openssl just need to have the config changed: https://www.openssl.org/docs/manmaster/man5/x509v3_config.html

Or can I simply add a switch to openssl for this option?

**Comment by FailedFace** [comment]
Looks like I need to follow this: https://access.redhat.com/solutions/28965
Which means my client is not generating a CSR that asks for 'extendedKeyUsage'.

I was thinking it was on the CRT creation not the CSR creation.

How bad is it to not have that extension for verifying the identity? I.E. Why does pykmip check?

**Comment by PeterHamilton** [comment]
The `extendedKeyUsage` extension is the "official" way for a certificate creator to indicate that the certificate should be used for client authentication. Technically a certificate without that extension should not be used to authenticate the user's identity. We added this check when we enhanced the authentication model for the PyKMIP server, since the server doesn't use other KMIP-based authentication methods (e.g., username/password, device credentials, attestation).

In the big scheme of things, you don't have to have the `extendedKeyUsage` extension. The PyKMIP client/server will work just fine without it, as long as the `enable_tls_client_auth` config setting is set to `False`.
